File: kotlin/lib/src/main/kotlin/desertroad/solar/Phase.kt
package desertroad.solar

import desertroad.solar.units.Angle

/**
 * Differentiation of horizon phases by altitude.
 */
enum class Phase {
    DAY,
    CIVIL_TWILIGHT,
    NAUTICAL_TWILIGHT,
    ASTRONOMICAL_TWILIGHT,
    NIGHT, ;

    companion object {
        internal fun fromAltitude(altitude: Angle) = when {
            altitude.inRadians >= Altitude.AT_SUNRISE.inRadians -> DAY
            altitude.inRadians >= Altitude.CIVIL.inRadians -> CIVIL_TWILIGHT
            altitude.inRadians >= Altitude.NAUTICAL.inRadians -> NAUTICAL_TWILIGHT
            altitude.inRadians >= Altitude.ASTRONOMICAL.inRadians -> ASTRONOMICAL_TWILIGHT
            else -> NIGHT
        }
    }
}

File: dart/lib/src/phase.dart
// Copyright 2023 Kim Heeyong
// MIT License

import 'package:local_solar/src/altitude.dart';
import 'package:local_solar/src/angle.dart';

/// Differentiation of horizon phases by altitude.
enum Phase {
  day,
  civilTwilight,
  nauticalTwilight,
  astronomicalTwilight,
  night, ;

  static Phase fromAltitude(Angle altitude) {
    if (altitude.inRadians >= Altitude.atSunrise.inRadians) {
      return Phase.day;
    } else if (altitude.inRadians >= Altitude.civil.inRadians) {
      return Phase.civilTwilight;
    } else if (altitude.inRadians >= Altitude.nautical.inRadians) {
      return Phase.nauticalTwilight;
    } else if (altitude.inRadians >= Altitude.astronomical.inRadians) {
      return Phase.astronomicalTwilight;
    } else {
      return Phase.night;
    }
  }
}